lib/travis/build/script/elm.rb

```diff
@@ -0,0 +1,127 @@
+module Travis
+  module Build
+    class Script
+      class Elm < NodeJs
+        # Default NodeJS version to install
+        DEFAULT_NODE_VERSION = '6.10.0'
+
+        DEFAULTS = {
+          elm_version: '0.18.0',
+          elm_test_version: '0.18.12'
+        }
+
+        ELM_TEST_REQUIRED_NODE_VERSION = '4.0.0'
+
+        def export
+          super
+          sh.export 'TRAVIS_ELM_VERSION', elm_version, echo: false
+          sh.export 'TRAVIS_ELM_TEST_VERSION', elm_test_version, echo: false
+        end
+
+        def configure
+          super
+
+          config[:node_js] ||= DEFAULT_NODE_VERSION
+
+          sh.if '! -d sysconfcpus/bin' do
+            install_sysconfcpus
+          end
+        end
+
+        def announce
+          super
+          sh.cmd 'elm --version', echo: true
+          sh.cmd 'elm-test --version', echo: true
+        end
+
+        def setup
+          super
+
+          install_elm
+          install_elm_test
+        end
+
+        def script
+          sh.cmd 'elm-format --validate .'
+          sh.cmd 'elm-test'
+        end
+
+        def setup_cache
+          if data.cache?(:elm)
+            sh.fold 'cache.elm' do
+              directory_cache.add '$HOME/.cache/elm-stuff', 'elm-stuff'
+
+              # some put their tests in ./test and others use ./tests
+              directory_cache.add '$HOME/.cache/elm-test-stuff', 'test/elm-stuff'
+              directory_cache.add '$HOME/.cache/elm-tests-stuff', 'tests/elm-stuff'
+
+              # we build sysconfcpus from source, so cache the result
+              directory_cache.add '$HOME/.cache/sysconfcpus-cache', 'sysconfcpus'
+            end
+          end
+        end
+
+        def cache_slug
+          super << '-elm-' << elm_version
+        end
+
+        private
+
+          def elm_version
+            config[:elm_version].to_s
+          end
+
+          def elm_test_version
+            config[:elm_test_version].to_s
+          end
+
+          def install_elm
```

Comment:
This portion turns out to be rather chatty (https://staging.travis-ci.org/BanzaiMan/travis_staging_test/jobs/711533#L1026-L1144). I suggest putting it in a fold.

Reply:
Put behind a fold in b6697ff5bbe8de4ef3a044fe312cdb5e56a9a384